pkg/plugins/secret/secret_manager/secret_manager.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2021 Nitric Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secret_manager_secret_service
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	gax "github.com/googleapis/gax-go/v2"
+	"github.com/nitric-dev/membrane/pkg/plugins"
+	"github.com/nitric-dev/membrane/pkg/plugins/secret"
+	"golang.org/x/oauth2/google"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type SecretManagerClient interface {
+	AccessSecretVersion(context.Context, *secretmanagerpb.AccessSecretVersionRequest, ...gax.CallOption) (*secretmanagerpb.AccessSecretVersionResponse, error)
+	AddSecretVersion(context.Context, *secretmanagerpb.AddSecretVersionRequest, ...gax.CallOption) (*secretmanagerpb.SecretVersion, error)
+	CreateSecret(context.Context, *secretmanagerpb.CreateSecretRequest, ...gax.CallOption) (*secretmanagerpb.Secret, error)
+	GetSecret(context.Context, *secretmanagerpb.GetSecretRequest, ...gax.CallOption) (*secretmanagerpb.Secret, error)
+	UpdateSecret(context.Context, *secretmanagerpb.UpdateSecretRequest, ...gax.CallOption) (*secretmanagerpb.Secret, error)
+}
+
+type secretManagerSecretService struct {
+	secret.UnimplementedSecretPlugin
+	client    SecretManagerClient
+	projectId string
+}
+
+func validateNewSecret(sec *secret.Secret, val []byte) error {
+	if sec == nil {
+		return plugins.NewInvalidArgError("secret must not be nil")
+	}
+	if len(sec.Name) == 0 {
+		return plugins.NewInvalidArgError("provide non-blank secret name")
+	}
+	if len(val) == 0 {
```

Comment:
We should probably add some common secret.Name validation around what a valid name is, e.g. text, numbers, dashes and . only for all the plugins.  Worried about / character in Name with version parsing later.

Reply:
will raise a new issue for this, as it will be common to all plugins (adapter layer validation).